chainer/datasets/pickle_dataset.py

```diff
@@ -0,0 +1,164 @@
+        self._protocol = protocol
+
+    def close(self):
+        self._writer.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def write(self, x):
+        position = self._writer.tell()
+        pickle.dump(x, self._writer, protocol=self._protocol)
+        self._positions.append(position)
+
+    def flush(self):
+        self._writer.flush()
+
+
+class PickleDataset(dataset_mixin.DatasetMixin):
+
+    """Dataset stored in a storage using pickle.
+
+    :mod:`pickle` is the default serialization library of Python.
+    This dataset stores any objects in a storage using :mod:`pickle`.
+    Even when a user want to use a large dataset, this dataset can stores all
+    data in a large storage like HDD and each data can be randomly accessible.
+
+    >>> with chainer.datasets.open_pickle_dataset_writer('/path/to/data') as w:
+    ...     w.write((1, 2.0, 'hello'))
+    ...     w.write((2, 3.0, 'good-bye'))
+    ...
+    >>> with chainer.datasets.open_pickle_dataset('/path/to/data') as dataset:
+    ...     print(dataset[1])
+    ...
+    (2, 3.0, 'good-bye')
+
+    Args:
+        reader: File like object. `reader` must support random access.
+
+    """
+
+    def __init__(self, reader):
+        # TODO(unnonouno): It is better to check seekable(), but the method is
+        # not supported in py2
+        self._reader = reader
+        self._positions = []
+        reader.seek(0)
+        while True:
+            position = reader.tell()
+            try:
+                pickle.load(reader)
+            except EOFError:
+                break
+            self._positions.append(position)
+
+        self._lock = threading.RLock()
+
+    def close(self):
+        """Closes a file reader.
+
+        After a user calls this method, the dataset is not accessible.
+        """
+        with self._lock:
+            self._reader.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def __len__(self):
+        return len(self._positions)
+
+    def get_example(self, index):
+        with self._lock:
+            self._reader.seek(self._positions[index])
+            return pickle.load(self._reader)
+
+
+def open_pickle_dataset(path):
+    """Opens a dataset stored in a given path.
+
+    This is a hepler funciton to open :class:`PickleDataset`. It opens a given
+    file in binary mode, and make :class:`PickleDataset` instance.
+
+    This method does not close the opened file. A user needs to call
+    :func:`PickleDataset.close`.
+
+    Args:
+        path (str): Path to a dataset.
+
+    Returns:
+        chainer.datasets.PickleDataset: Opened dataset.
+
+    .. seealso: chainer.datasets.PickleDataset
+
+    """
+    reader = open(path, 'rb')
+    try:
+        return PickleDataset(reader)
+    except Exception:
+        try:
+            reader.close()
```

Comment:
I thought we do not need to explicitly close `reader` here because anyway there will be no reference to `reader` in this path and the file will be closed.

Reply:
Hum. So you mean it is ok to believe reference counter, right?

Reply:
Yes.